sound/soc/sof/intel/mtl.c

```diff
@@ -0,0 +1,941 @@
+	}
+
+	dev_dbg(sdev->dev, "Primary core power up successful\n");
+
+	/* step 4: wait for IPC DONE bit from ROM */
+	ret = snd_sof_dsp_read_poll_timeout(sdev, HDA_DSP_BAR, chip->ipc_ack, status,
+					    ((status & chip->ipc_ack_mask) == chip->ipc_ack_mask),
+					    HDA_DSP_REG_POLL_INTERVAL_US, MTL_DSP_PURGE_TIMEOUT_US);
+	if (ret < 0) {
+		if (hda->boot_iteration == HDA_FW_BOOT_ATTEMPTS)
+			dev_err(sdev->dev, "timeout waiting for purge IPC done\n");
+		goto err;
+	}
+
+	/* set DONE bit to clear the reply IPC message */
+	snd_sof_dsp_update_bits_forced(sdev, HDA_DSP_BAR, chip->ipc_ack, chip->ipc_ack_mask,
+				       chip->ipc_ack_mask);
+
+	/* step 5: enable interrupts */
+	ret = mtl_enable_interrupts(sdev);
+	if (ret < 0) {
+		if (hda->boot_iteration == HDA_FW_BOOT_ATTEMPTS)
+			dev_err(sdev->dev, "%s: failed to enable interrupts\n", __func__);
+		return ret;
+	}
+
+	mtl_enable_ipc_interrupts(sdev);
+
+	/* step 6: wait for ROM_INIT_DONE */
+	/*
+	 * ACE workaround as in Windows driver.
+	 * "ACE platform cannot catch FSR_ROM_INIT_DONE, because this status is setting by short
+	 * time. This "for" loop is redundant for us and we get timeout almost everytime by this
+	 * condition."
+	 * So dont wait for ROM INIT.
+	 */
+
+	return 0;
+
+err:
+	snd_sof_dsp_dbg_dump(sdev, "MTL DSP init fail", 0);
+	mtl_dsp_core_power_down(sdev, chip->host_managed_cores_mask);
+	return ret;
+}
+
+static int mtl_dsp_boot_imr(struct snd_sof_dev *sdev)
+{
+	int ret;
+
+	ret = mtl_dsp_cl_init(sdev, 0, true);
+	if (ret >= 0)
+		hda_sdw_process_wakeen(sdev);
+
+	return ret;
+}
+
+static int mtl_dsp_cl_boot_firmware(struct snd_sof_dev *sdev)
+{
+	struct sof_intel_hda_dev *hda = sdev->pdata->hw_pdata;
+	struct snd_sof_pdata *plat_data = sdev->pdata;
+	const struct sof_dev_desc *desc = plat_data->desc;
+	const struct sof_intel_dsp_desc *chip_info;
+	struct firmware stripped_firmware;
+	struct hdac_ext_stream *hext_stream;
+	struct snd_dma_buffer dmab;
+	int ret1;
+	int ret;
+	int i;
+
+	if (hda->imrboot_supported && !sdev->first_boot) {
+                dev_dbg(sdev->dev, "IMR restore supported, booting from IMR directly\n");
+                hda->boot_iteration = 0;
+                ret = mtl_dsp_boot_imr(sdev);
+                if (ret >= 0)
+                        return ret;
+
+                dev_warn(sdev->dev, "IMR restore failed, trying to cold boot\n");
+        }
+
+	chip_info = desc->chip_info;
+
+	if (plat_data->fw->size <= plat_data->fw_offset) {
+		dev_err(sdev->dev, "firmware size must be greater than firmware offset\n");
+		return -EINVAL;
+	}
+
+	stripped_firmware.data = plat_data->fw->data + plat_data->fw_offset;
+	stripped_firmware.size = plat_data->fw->size - plat_data->fw_offset;
+
+	/* init for booting wait */
+	init_waitqueue_head(&sdev->boot_wait);
+
+	/* prepare DMA for code loader stream */
+	hext_stream = hda_cl_stream_prepare(sdev, HDA_CL_STREAM_FORMAT,
+					    stripped_firmware.size,
+					    &dmab, SNDRV_PCM_STREAM_PLAYBACK);
```

Comment:
alignment?

Reply:
> alignment?

The alignment looks good to me, IMHO.

Reply:
The code will look like below, it is good, right?
```
        hext_stream = hda_cl_stream_prepare(sdev, HDA_CL_STREAM_FORMAT,
                                            stripped_firmware.size,
                                            &dmab, SNDRV_PCM_STREAM_PLAYBACK);
```

Reply:
yep!